pfcpiface/conn.go

```diff
@@ -63,6 +65,80 @@ type PFCPConn struct {
 	shutdown chan struct{}
 
 	metrics.InstrumentPFCP
+	hbMonitor   *hbMonitor
+	pendingReqs sync.Map
+}
+
+type HbTimerStatus uint8
+
+const (
+	CancelHbTimer HbTimerStatus = iota
+	StartHbTimer
+	StopHbTimer
+)
+
+type hbMonitor struct {
+	hbMonCh chan HbTimerStatus
+}
+
+func (pConn *PFCPConn) startHeartBeatMonitor() {
+	// Stop HeartBeat routine if already running
+	pConn.CancelHeartBeatTimer()
+
+	if !pConn.upf.enableHBTimer {
+		return
+	}
+
+	hbCh := make(chan HbTimerStatus)
+	pConn.hbMonitor = &hbMonitor{hbMonCh: hbCh}
+
+	go func() {
+		log.Infoln("Starting Heartbeat timer")
+		heartBeatTimer := time.NewTimer(pConn.upf.hbInterval)
+
+		for {
+			select {
+			case <-pConn.ctx.Done():
+				heartBeatTimer.Stop()
+				return
+
+			case status := <-hbCh:
+
+				if status == StartHbTimer {
+					log.Println("Start HeartBeat Timer")
+					// Start the hb interval timer
+					heartBeatTimer = time.NewTimer(pConn.upf.hbInterval)
+				} else if status == CancelHbTimer {
+					log.Println("Cancel HeartBeat Timer")
+					return
+				}
+
+			case <-heartBeatTimer.C:
+				log.Println("HeartBeat Interval Timer Expired")
+
+				r := pConn.sendHeartBeatRequest()
+
+				pConn.WaitForResponse(r, pConn.upf.maxReqRetries,
+					func(msg message.Message) bool {
+						pConn.Shutdown()
+						return true
+					})
```

Comment:
Can we simplify by only calling GetResponse on the heartbeat request? We can remove any response handling code elsewhere.

Reply:
the invoker will get back the full reply so no need for GetCause

Reply:
Should the retries be part of the GetResponse API? In that case, we need to pass the pfcp conn information as well.

Reply:
Can get rid of StartHeartBeatTimer

Reply:
@clstefi that should be fine GetResponse(pConn.retries, pConn.Timeout)

Reply:
based on offline discussion, guess we need WaitForResponse as that is on pConn and the only way to resend requests